compiler/rustc_typeck/src/astconv/mod.rs

```diff
@@ -1318,7 +1319,8 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
 
         // De-duplicate auto traits so that, e.g., `dyn Trait + Send + Send` is the same as
         // `dyn Trait + Send`.
-        auto_traits.sort_by_key(|i| i.trait_ref().def_id());
+        // Sort by `DefPathHash` so that the order is stable across compilation sessions
+        auto_traits.sort_by_key(|i| self.tcx().def_path_hash(i.trait_ref().def_id()));
         auto_traits.dedup_by_key(|i| i.trait_ref().def_id());
```

Comment:
This is probably the reason for the test failures: `dedup_by_key` assumes that the `Vec` is sorted by the given key. So this should be `auto_traits.dedup_by_key(|i| self.tcx().def_path_hash(i.trait_ref().def_id()));`

Maybe it would make sense to use a `rustc_data_structures::sorted_map::SortedMap` here.

Reply:
Since def path hash equality implies that the `DefId`s are equal, I thought that deduplicating by `DefId` would still work, since equal `DefIds` should still be adjacent to each other.

Reply:
Also, the failure appears to be caused be re-ordered errors, not duplicated errors.